lib/arel/transformer/replace_table_with_subquery.rb

```diff
@@ -0,0 +1,31 @@
+module Arel
+  module Transformer
+    class ReplaceTableWithSubquery
+      attr_reader :subquery_for_table
+
+      def initialize(subquery_for_table)
+        @subquery_for_table = subquery_for_table
+      end
+
+      def call(arel, next_middleware)
+        tree = Arel.enhance(arel)
+        update_arel_tables(tree)
+        next_middleware.call tree
+      end
+
+      private
+
+      def update_arel_tables(tree)
+        tree.query(
+          class: Arel::Table,
+          context: { range_variable: true },
+          schema_name: nil,
+        ).each do |node|
+          if (subquery = subquery_for_table.call(node.name.value))
```

Comment:
@mvgijssel I've changed this strategy a bit, rather than passing a (frozen) mapping you have to pass a callback now.

Reply:
Looks good!